### PR TITLE
Fix infinite loop for convolution with extent equal 1

### LINKED ISSUE #3277: Gaussian Blur for grayscale images leads to stuck on GPU
Hi! I try to perform gaussian blur for grayscale images (channels = 1) and faced a problem that in this case pipeline is stuck and GPU usage bounces up to 100%. Observe this behavior both for CUDA 10.2 and CUDA 11.0 versions (DALI 1.5.0), also on different hardware: MX150, RTX2080Ti. Code to reproduce:
```
import numpy as np
import nvidia.dali as dali


DEVICE_ID = {"cpu": dali.types.CPU_ONLY_DEVICE_ID, "gpu": 0}
DEVICE = "gpu"  # If DEVICE = "cpu" it works with any CHANNELS value
BATCH_SIZE = 1
HEIGHT = 128
WIDTH = 128
CHANNELS = 1  # If CHANNELS >= 2 it works, but DEVICE = "gpu" and CHANNELS = 1 caused stuck


def data_injection():
    inputs = np.random.random_integers(low=0, high=255, size=(BATCH_SIZE, HEIGHT, WIDTH, CHANNELS)).astype(np.uint8)
    targets = np.random.random_integers(low=0, high=255, size=(BATCH_SIZE, HEIGHT, WIDTH, CHANNELS)).astype(np.uint8)
    return inputs, targets


augmentation_pipeline = dali.Pipeline(batch_size=BATCH_SIZE,
                                      num_threads=4,
                                      device_id=DEVICE_ID[DEVICE],
                                      exec_async=False, exec_pipelined=False, seed=42)
with augmentation_pipeline:
    inputs, targets = dali.fn.external_source(source=data_injection, device=DEVICE, num_outputs=2)
    blurred_targets = dali.fn.gaussian_blur(targets, device=DEVICE, sigma=2.0)
    augmentation_pipeline.set_outputs(inputs, blurred_targets)

augmentation_pipeline.build()

pipeline_result = augmentation_pipeline.run()
pipeline_result = pipeline_result if DEVICE == "cpu" else list(map(lambda item: item.as_cpu(), pipeline_result))
print(pipeline_result[0].as_array().shape, pipeline_result[1].as_array().shape)
```

### PR DESCRIPTION
**Pull request by klecki**
## Description
- [x] **Bug fix** (*non-breaking change which fixes an issue*)
- [ ] **New feature** (*non-breaking change which adds functionality*)
- [ ] **Breaking change** (*fix or feature that would cause existing functionality to not work as expected*)
- [ ] **Refactoring** (*Redesign of existing code that doesn't affect functionality*)
- [ ] **Other** (*e.g. Documentation, Tests, Configuration*)

### What happened in this PR
GPU kernel's border handling went into infinite loop,
as the border reflect 101 is not suited for inputs
with extent zero. As the CPU version treats it as
border replicate, we do the same for GPU.

As it boils down to `sum(window) * input_element`
we compact the window to 1 element and break the
loop early.

Test to cover edge cases added.

Fixes #3277.

#### Additional information
- Affected modules and functionalities:
Convolution/GPU

- Key points relevant for the review:

### Tests
- [ ] Existing tests apply
- [x] New tests added
  - [x] Python tests
  - [x] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [x] Existing documentation applies
- [ ] Documentation updated
  - [ ] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [ ] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [x] N/A

**REQ IDs**: N/A

**JIRA TASK**: N/A

### CODE REVIEW

File: dali/kernels/imgproc/convolution/convolution_gpu.h [hunk 1]
```diff
@@ -236,17 +246,45 @@ struct ConvolutionGpu {
...
+      // (sum(window) * input_element). The kernel matrix generation ends as infinite loop
+      // for input of extent equal 1, so we compact the kernel to 1 element (what would effectively
+      // happen either way) and just lookup that one elemnt in the CUTLASS kernels
+      if (in_shape[i][axis] == 1) {
+        std::array<W, 1> tmp_window = {W{0}};
+        for (int win_elem = 0; win_elem < window.tensor_shape_span(i)[0] / 2; win_elem++) {
+          tmp_window[0] += window.tensor_data(i)[win_elem];
+          tmp_window[0] += window.tensor_data(i)[window.tensor_shape_span(i)[0] - win_elem - 1];
+        }
+        tmp_window[0] += window.tensor_data(i)[window.tensor_shape_span(i)[0] / 2];
```
Comment (mzient) [inline dali/kernels/imgproc/convolution/convolution_gpu.h:281]: Why not just a plain loop???

Comment (mzient) [inline dali/kernels/imgproc/convolution/convolution_gpu.h:281]: Why not just a plain loop?
```suggestion
        for (int win_elem = 0; win_elem < window.tensor_shape_span(i)[0]; win_elem++) {
          tmp_window[0] += window.tensor_data(i)[win_elem];
        }
```

File: dali/test/python/test_operator_gaussian_blur.py [hunk 2]
```diff
@@ -223,6 +228,11 @@ def test_generic_gaussian_blur():
         for (t_in, t_out) in [(np.uint8, types.NO_TYPE), (np.float32, types.FLOAT), (np.uint8, types.FLOAT)]:
             yield from generate_generic_cases(dev, t_in, t_out)
 
+def test_one_sized_extent():
+    for dev in ["cpu", "gpu"]:
+        for shape, layout in [((10, 1, 3), "HWC"), ((1, 10, 3), "HWC"), ((1, 10), "HW"), ((10, 1), "HW")]:
+            yield check_generic_gaussian_blur, 10, 2.0, 5, shape, layout, 2, dev, np.float32, types.FLOAT, False
```
Comment (mzient) [inline dali/test/python/test_operator_gaussian_blur.py:234]: Add a 3D test without channels, with the outermost extent of 1.

File: dali/kernels/imgproc/convolution/convolution_gpu.h (hunk 1 above)
Comment (mzient) [reply dali/kernels/imgproc/convolution/convolution_gpu.h:281]: Use Kahan's sum instead?

File: dali/kernels/imgproc/convolution/cutlass/threadblock/predicated_tile_iterator.h [hunk 3]
```diff
@@ -214,7 +214,7 @@ namespace threadblock {
 /// Same when going to the bottom, 1 + 2 * 6 = 13 - also out.
 /// Note that, when the distance is 0, we don't add the additional coefficients in this border mode.
 ///
-/// Make channel number > 1 will cause additional gaps in the kernel when looking at the
+/// Makeing the channel number > 1 will cause additional gaps in the kernel when looking at the
```
Comment (JanuszL) [inline dali/kernels/imgproc/convolution/cutlass/threadblock/predicated_tile_iterator.h:217]:
```suggestion
/// Making the channel number > 1 will cause additional gaps in the kernel when looking at the
```

File: dali/kernels/imgproc/convolution/convolution_gpu.h (hunk 1 above)
Comment (klecki) [reply dali/kernels/imgproc/convolution/convolution_gpu.h:281]: done

File: dali/kernels/imgproc/convolution/cutlass/threadblock/predicated_tile_iterator.h (hunk 3 above)
Comment (klecki) [reply dali/kernels/imgproc/convolution/cutlass/threadblock/predicated_tile_iterator.h:217]: done

File: dali/test/python/test_operator_gaussian_blur.py (hunk 2 above)
Comment (klecki) [reply dali/test/python/test_operator_gaussian_blur.py:234]: Should be equivalent to the 2D case with non-innermost extent equal one, but I can add one just in case.

**Review by JanuszL — APPROVED** [review]

**Comment by klecki** [comment]
!build

**Comment by klecki** [comment]
!build

**Review by mzient — APPROVED** [review]
